File: src/LeetCode/Algorithms/Java/Contiguous_Array.java
package LeetCode.Algorithms.Java;

import java.util.Hashtable;

// https://leetcode.com/problems/contiguous-array/
public class Contiguous_Array {
	public static void main(String[] args) {
		int[] nums = {0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1}; // {0, 1};
		
		System.out.println(findMaxLength(nums));
	}
	
	public static int findMaxLength(int[] nums) {
		int result = 0, count = 0;
		Hashtable<Integer, Integer> hashtable = new Hashtable<>();
		
		hashtable.put(0, -1);
		
		for (int i = 0; i < nums.length; i++) {
			count += (nums[i] == 1) ? 1 : -1;
			
			if (hashtable.containsKey(count)) {
				result = Math.max(result, i - hashtable.get(count));
			} else {
				hashtable.put(count, i);
			}
		}
		
		return result;
	}
}

File: src/LeetCode/Algorithms/JavaScript/Contiguous_Array.js

/**
 * https://leetcode.com/problems/contiguous-array/solution/
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function(nums) {
	const hashMap = new Map();
	let maxlen = 0;
	let count = 0;
	
	hashMap.set(0, -1);
	
	for (let i = 0; i < nums.length; i++) {
		count = count + (nums[i] === 1 ? 1 : -1);
		
		if (hashMap.has(count)) {
			maxlen = Math.max(maxlen, i - hashMap.get(count));
		} else {
			hashMap.set(count, i);
		}
	}
	
	return maxlen;
};

// const nums = [0, 1];
const nums = [0, 0, 1];

console.log(findMaxLength(nums));